src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs

```diff
@@ -190,108 +193,132 @@ private TestRunResult BuildTestRunResult(IRunResults testResults, int expectedTe
-                    return eventHandler;
-                }
-                _logger.LogWarning($"{RunnerId}: Retrying the test session.");
+                _logger.LogWarning(
+                    $"{RunnerId}: Test assembl{(sources.Count() < 2 ? "y does" : "ies do")} not contain any test, skipping.");
+                return (new SimpleRunResults(), new SimpleRunResults());
             }
 
-            _logger.LogCritical($"{RunnerId}: VsTest failed, settings: {runSettings}");
-
-            throw new GeneralStrykerException(
-                $"{RunnerId}: failed to run a test session despite ${MaxAttempts} attempts. Aborting session.");
-        }
-
-        private RunEventHandler RunVsTest(ITestGuids tests, IEnumerable<string> sources, string runSettings, int? timeOut,
-            Action<IRunResults> updateHandler)
-        {
-            var eventHandler = new RunEventHandler(_context.VsTests, _logger, RunnerId);
-            var vsTestFailed = false;
-            var sessionFailed = false;
-
-            void HandlerVsTestFailed(object sender, EventArgs e)
-            {
-                sessionFailed = true;
-            }
+            var runEventHandler = new RunEventHandler(_context.VsTests, _logger, RunnerId);
 
             void HandlerUpdate(object sender, EventArgs e)
             {
-                updateHandler?.Invoke(eventHandler.GetResults());
+                updateHandler?.Invoke(runEventHandler.GetResults());
             }
 
-            var strykerVsTestHostLauncher = _context.BuildHostLauncher(RunnerId);
-
-            eventHandler.VsTestFailed += HandlerVsTestFailed;
-            eventHandler.ResultsUpdated += HandlerUpdate;
-
-            _cancelled = false;
             var options = new TestPlatformOptions
             {
                 TestCaseFilter = string.IsNullOrWhiteSpace(_context.Options.TestCaseFilter)
                     ? null
                     : _context.Options.TestCaseFilter
             };
-            Task session;
-            if (tests.IsEveryTest)
+
+            runEventHandler.ResultsUpdated += HandlerUpdate;
+            // work around VsTest issues when using multiple test assemblies
+            foreach (var source in validSources)
             {
-                session = _vsTestConsole.RunTestsWithCustomTestHostAsync(sources, runSettings, options, eventHandler,
-                    strykerVsTestHostLauncher);
+                var testForSource = _context.TestsPerSource[source];
+                var testsForAssembly = new TestGuidsList(tests.GetGuids()?.Where(testForSource.Contains));
+                if (!tests.IsEveryTest && testsForAssembly.Count == 0)
+                {
+                    // skip empty assemblies
+                    continue;
+                }
+
+                _logger.LogTrace($"{RunnerId}: testing assembly {source}.");
+                RunVsTest(tests, source, runSettings, options, timeOut, runEventHandler);
+
+                if (_cancelled)
```

Comment:
VSTestRunner is still mostly dark matter for me - yet I wonder, have you considered bringing `CancellationToken` here for a more native handling of cancellations? That could probably simplify testing as well.

Reply:
I guess this is poor naming on my part. `_cancelled` is true when Stryker bails the test session because it reaches a conclusion for every mutation to be tested (in this session). It is used here so that Stryker does not start a test session with the next assembly in the list. It saves time (and avoid a few nasty bugs linked to extra notifications).

Reply:
Ohhh I see. Maybe `_testSessionCanceled` then?

Reply:
went for `_currentSessionCancelled'